gh-113884: Make queue.SimpleQueue thread-safe in --disable-gil builds

mpage opened the pull request:
Make `queue.SimpleQueue` thread-safe when the GIL is disabled. Queue state is protected by the per-object lock; thread suspension now uses the low-level `PyParkingLot` abstraction.

<!-- gh-issue-number: gh-113884 -->
* Issue: gh-113884
<!-- /gh-issue-number -->

mpage committed:
Use ParkingLot to manage waiting threads

mpage committed:
Make SimpleQueue thread-safe with the GIL disabled

Methods on SimpleQueue are protected with the per-object lock.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

erlend-aasland commented on Misc/NEWS.d/next/Core and Builtins/2024-01-17-00-52-57.gh-issue-113884.CvEjUE.rst:
```diff
@@ -0,0 +1 @@
+Make ``queue.SimpleQueue`` thread safe when the GIL is disabled. Existing semantics and behavior should be preserved, but the implementation is changed to ensure thread-safety when the GIL is disabled.
```
```suggestion
Make :class:`queue.SimpleQueue` thread safe when the GIL is disabled. Existing semantics and behavior should be preserved, but the implementation is changed to ensure thread-safety when the GIL is disabled.
```

erlend-aasland committed:
Update Misc/NEWS.d/next/Core and Builtins/2024-01-17-00-52-57.gh-issue-113884.CvEjUE.rst

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -25,12 +26,158 @@ static struct PyModuleDef queuemodule;
 #define simplequeue_get_state_by_type(type) \
     (simplequeue_get_state(PyType_GetModuleByDef(type, &queuemodule)))
 
+static const Py_ssize_t kInitialRingBufCapacity = 8;
```
The prevailing style in CPython seems to be `INITIAL_RING_BUF_CAPACITY` instead of `kInitialRingBufCapacity`.

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -231,66 +370,71 @@ _queue_SimpleQueue_get_impl(simplequeueobject *self, PyTypeObject *cls,
-        microseconds = _PyTime_AsMicroseconds(timeout,
-                                              _PyTime_ROUND_CEILING);
+        PY_TIMEOUT_T microseconds =
+            _PyTime_AsMicroseconds(timeout, _PyTime_ROUND_CEILING);
```
The microseconds overflow check is no longer relevant because we are not using `PyThread_acquire_lock_timed()`.

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -25,12 +26,158 @@ static struct PyModuleDef queuemodule;
-    Py_ssize_t lst_pos;
+
+    // Are there threads waiting for items
+    uint8_t threads_waiting;
```
Maybe `bool threads_waiting` or `bool has_threads_waiting`?

Otherwise, it's a bit unclear if this is a count of waiting threads.

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -25,12 +26,158 @@ static struct PyModuleDef queuemodule;
+    Py_ssize_t new_capacity = Py_MAX(kInitialRingBufCapacity, buf->items_cap * 2);
+    PyObject **new_items = PyMem_Calloc(new_capacity, sizeof(PyObject *));
+    if (new_items == NULL) {
+        return -1;
```
This needs a `PyErr_NoMemory()` call either here or further up the call chain.

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -25,12 +26,158 @@ static struct PyModuleDef queuemodule;
+    buf->put_idx = 0;
+    buf->get_idx = 0;
+    buf->items_cap = kInitialRingBufCapacity;
+    buf->items = PyMem_Calloc(buf->items_cap, sizeof(PyObject*));
```
This needs a PyErr_NoMemory() call on failure either here or further up the call chain.

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -25,12 +26,158 @@ static struct PyModuleDef queuemodule;
+    if (buf->num_items < (buf->items_cap / 2)) {
+        // Items is less than 50% occupied, shrink it
+        shrink_ringbuf(buf);
+    }
```
If we double in size when growing then I think the threshold for shrinking should be smaller, (like 25% occupied). Otherwise we can end up with quadratic behavior from adding and removing one or two elements near the capacity -- shrinking by half at 50% means that we are at almost 100% capacity.

mpage replied:
Yeah I was concerned about this too and was chatting with @DinoV about whether or not we should use a lower threshold here. I opted not to because I believe this matches the behavior of the previous implementation. Happy to switch to 25%.

colesbury replied:
This matches the list shrinking behavior, but list growth is slower. For large sizes it grows by 1.125x instead of 2x, so it doesn't risk the quadratic behavior.

colesbury reviewed (commented):
Other than the threshold for shrinking and a few minor comments, this looks good to me.

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -231,66 +370,71 @@ _queue_SimpleQueue_get_impl(simplequeueobject *self, PyTypeObject *cls,
+                break;
+            }
+            case Py_PARK_AGAIN: {
+                // This should be impossible with the current implementation of
+                // PyParkingLot, but would be possible if critical sections /
+                // the GIL were released before the thread was added to the
+                // internal thread queue in the parking lot.
```
I agree that this shouldn't happen and also this looks like the correct behavior (loop again) in case it ever did happen in the future.

mpage committed:
Remove unnecessary overflow check

We're not using `PyThread_acquire_lock_timed`.

mpage committed:
Use a clearer name for field indicating whether or not threads are waiting

colesbury commented on Modules/_queuemodule.c:
```diff
@@ -361,65 +386,64 @@ _queue_SimpleQueue_get_impl(simplequeueobject *self, PyTypeObject *cls,
-    if (self->locked) {
-        PyThread_release_lock(self->lock);
-        self->locked = 0;
+        uint8_t waiting = 1;
```
Should be `bool waiting = true` now

colesbury reviewed (commented):
@mpage -  you probably want to mark the PR as "Ready for review" (i.e., not a "Draft")

This LGTM other than one minor comment.

mpage committed:
Fix type

erlend-aasland commented on Modules/_queuemodule.c:
```diff
@@ -263,8 +258,29 @@ simplequeue_new_impl(PyTypeObject *type)
 
     return (PyObject *) self;
 }
+typedef struct {
```
Style nit.

```suggestion

typedef struct {
```

erlend-aasland commented on Modules/_queuemodule.c:
```diff
@@ -335,23 +368,15 @@ in that case).
-        microseconds = 0;
-    }
-    else if (timeout_obj != Py_None) {
+    if (block != 0 && timeout_obj != Py_None) {
```
```suggestion
    if (block != 0 && !Py_IsNone) {
```

erlend-aasland commented on Modules/_queuemodule.c:
```diff
@@ -307,12 +330,22 @@ for compatibility with the Queue class.
+static PyObject *
+empty_error(PyTypeObject *cls)
+{
+    PyObject *module = PyType_GetModule(cls);
```
```suggestion
    PyObject *module = PyType_GetModule(cls);
    assert(module != NULL);
```

erlend-aasland commented on Misc/NEWS.d/next/Core and Builtins/2024-01-17-00-52-57.gh-issue-113884.CvEjUE.rst:
```diff
@@ -0,0 +1 @@
+Make :class:`queue.SimpleQueue` thread safe when the GIL is disabled. Existing semantics and behavior should be preserved, but the implementation is changed to ensure thread-safety when the GIL is disabled.
```
```suggestion
Make :class:`queue.SimpleQueue` thread safe when the GIL is disabled.
```

erlend-aasland reviewed (commented):
Thanks. The parking lot API still feels little bit uncomfortable for me, but AFAICS, this looks good. Perhaps a comment explaining how the hand-off data structure relates to the parking lot API could be beneficial for future readers of the code. I also find the `maybe_unparked_thread` oddly named. OTOH, I don't have a better suggestion :)

erlend-aasland commented on Modules/_queuemodule.c:
```diff
@@ -263,8 +258,29 @@ simplequeue_new_impl(PyTypeObject *type)
+        data->handed_off = 1;
+    }
+    data->queue->has_threads_waiting = has_more_waiters;
+}
```
I think what bothers me about the naming here, is that it is focused on the possible waiting thread. Perhaps it would be clearer to name this based on the actual queue operation?

Also, perhaps also consider using a `bool` datatype for `HandoffData.handed_off` (nit).

Feel free to ignore this, if you think I'm totally wrong :)

```suggestion
check_data_handoff(HandoffData *data, PyObject **item, bool has_more_waiters)
{
    // Check if data was handed off to a waiting thread.
    if (item == NULL) {
        data->handed_off = false;
    }
    else {
        *item = data->item;
        data->handed_off = true;
    }
    data->queue->has_threads_waiting = has_more_waiters;
}
```

mpage replied:
> Perhaps it would be clearer to name this based on the actual queue operation?

This looks good to me! 

> Also, perhaps also consider using a bool datatype for HandoffData.handed_off (nit).

Yeah, this is better. I'm still not fully used to using `bool` in C :)

colesbury replied:
The `has_more_waiters` function parameter should be an `int`. It needs to match the signature of `_Py_unpark_fn_t`. (We could change that signature to use a `bool`, but maybe in a separate PR?)

erlend-aasland replied:
> I'm still not fully used to using bool in C :)

Me neither 😄

> The `has_more_waiters` function parameter should be an `int`. It needs to match the signature of `_Py_unpark_fn_t`. (We could change that signature to use a `bool`, but maybe in a separate PR?)

For file scope, `stdbool.h` stuff is ok, but we don't want it to accidentally leak out through Python.h again[^1]. Let's be conservative and keep the parking lot API on `int`.


[^1]: #90904

erlend-aasland committed:
Apply suggestions from code review

erlend-aasland commented on Modules/_queuemodule.c:
```diff
@@ -335,23 +370,15 @@ in that case).
-        microseconds = 0;
-    }
-    else if (timeout_obj != Py_None) {
+    if (block != 0 && !Py_IsNone) {
```
```suggestion
    if (block != 0 && !Py_IsNone(timeout_obj)) {
```

erlend-aasland committed:
Fix suggestion

mpage committed:
Address review comments

Rename parking lot callback to better reflect what it does.
More precise type for whether handoff occurred.

erlend-aasland reviewed (approved):
Thank you so much, this is very nice!

kumaraditya303 commented on Modules/_queuemodule.c:
```diff
@@ -164,7 +167,7 @@ RingBuf_Put(RingBuf *buf, PyObject *item)
             return -1;
```
You need to decref item here isn't it?

colesbury replied:
Yes, I think so
